lib/policy_machine_storage_adapters/active_record/postgresql.rb

```diff
@@ -2,6 +2,127 @@
 
 module PolicyMachineStorageAdapter
   class ActiveRecord
+    class PolicyElementAssociation
+      def self.all_accessible_objects(associations, root_id:, filters: {})
+        pea_ids = associations.pluck(:id)
+        return [] if pea_ids.empty?
+
+        query = <<-SQL
+          id IN (
+            WITH RECURSIVE ancestor_scope AS (
+            SELECT
+                id AS child_id,
+                id AS parent_id
+            FROM policy_elements
+            WHERE id = ?
+
+            UNION ALL
+
+            SELECT
+                assignments.child_id,
+                assignments.parent_id
+            FROM assignments
+            JOIN ancestor_scope
+                ON assignments.child_id = ancestor_scope.parent_id
+            ),
+
+            leaf_ancestors AS (
+            SELECT DISTINCT
+                ancestor_scope.parent_id AS parent_id
+            FROM ancestor_scope
+            LEFT OUTER JOIN assignments
```

Comment:
Does this need to be a left outer join? I know we've seen performance issues around left outers that we can avoid before.

Reply:
What do you suggest as an alternative? I can't really think of a different way to do it.

Reply:
Talked offline. I misread this. It's fine as is 👍 .